agent/agent.go

```diff
@@ -564,12 +566,116 @@ func (o *opAgent) Transfer(stream m3em.Operator_TransferServer) error {
 		return err
 	}
 
-	return stream.SendAndClose(&m3em.TransferResponse{
+	return stream.SendAndClose(&m3em.PushFileResponse{
 		FileChecksum:   checksum.Current(),
 		NumChunksRecvd: int32(numChunks),
 	})
 }
 
+func validatePullFileRequest(request *m3em.PullFileRequest) error {
+	if request == nil {
+		return grpc.Errorf(codes.InvalidArgument, "nil request")
+	}
+
+	if request.ChunkSize <= 0 {
+		return grpc.Errorf(codes.InvalidArgument, "chunkSize must be a positive integer")
+	}
+
+	if request.MaxSize < 0 {
+		return grpc.Errorf(codes.InvalidArgument, "maxSize must be a non-negative integer")
+	}
+
+	return nil
+}
+
+// PullFile sends a local agent file to the caller
+func (o *opAgent) PullFile(request *m3em.PullFileRequest, stream m3em.Operator_PullFileServer) error {
+	if err := validatePullFileRequest(request); err != nil {
+		return err
+	}
+	o.logger.Infof("received PullFile(): %+v", *request)
+
+	o.RLock()
+	defer o.RUnlock()
+
+	if !o.isSetupWithLock() {
+		return grpc.Errorf(codes.InvalidArgument, "agent has not been setup, unable to transfer file")
+	}
+
+	pm := o.processMonitor
+	if pm == nil {
+		return grpc.Errorf(codes.InvalidArgument, "no process running, unable to transfer file")
+	}
+
+	switch fileType := request.GetFileType(); fileType {
+	case m3em.PullFileType_PULL_FILE_TYPE_SERVICE_STDERR:
+		return o.sendLocalFileWithRLock(pm.StderrPath(), request.ChunkSize, request.MaxSize, stream)
+
+	case m3em.PullFileType_PULL_FILE_TYPE_SERVICE_STDOUT:
+		return o.sendLocalFileWithRLock(pm.StdoutPath(), request.ChunkSize, request.MaxSize, stream)
+
+	default:
+		return grpc.Errorf(codes.InvalidArgument, "received unknown pull file: %v", fileType)
+	}
+}
+
+func (o *opAgent) sendLocalFileWithRLock(localPath string, chunkSize int64, maxBytes int64, stream m3em.Operator_PullFileServer) error {
+	fi, err := os.Stat(localPath)
+	if err != nil {
+		return grpc.Errorf(codes.InvalidArgument, "unable to find file: %v", err)
+	}
+
+	fd, err := os.Open(localPath)
+	if err != nil {
+		return grpc.Errorf(codes.InvalidArgument, "unable to open file: %v", err)
+	}
+
+	var (
+		reader    = bufio.NewReaderSize(fd, int(chunkSize))
+		buf       = make([]byte, chunkSize)
+		chunkIdx  = 1
+		truncated = false
+	)
+
+	// check if we need to seek ahead or if we are sending all the bytes
+	if maxBytes > 0 && fi.Size() > maxBytes {
+		offset := fi.Size() - maxBytes
+		if _, err := fd.Seek(offset, 0 /* relative to start of file */); err != nil {
+			return grpc.Errorf(codes.Internal, "unable to seek file: %v", err)
+		}
+		truncated = true
+	}
+
+	for {
+		n, err := reader.Read(buf)
+		switch err {
+		case io.EOF:
+			// i.e. streamed through the file, we can indicate we're done
```

Comment:
Just FYI - I've run into the case where `err == io.EOF` but `n > 0`.  For that case I've done this:

```go
eof := false 
for !eof {
  n, err := reader.Read(buf)
  if err == io.EOF {
    eof = true
  } else if err != nil {
    return fmt.Errorf("...: %v", err) // wrapped err
  }

  // use buf[:n]
}
```